association_backend/association/domain/repositories/association_repository.py

```diff
@@ -0,0 +1,121 @@
+import logging
+from typing import Optional
+
+import stripe
+from association.domain.entities.stripe_entities import (
+    StripeCheckoutSession,
+    StripeCheckoutSessionInput,
+    StripeCustomer,
+)
+from association.domain.entities.subscription_entities import Subscription
+from association.domain.repositories.base import AbstractRepository
+from association.settings import (
+    DOMAIN_URL,
+    STRIPE_SUBSCRIPTION_CANCEL_URL,
+    STRIPE_SUBSCRIPTION_PRICE_ID,
+    STRIPE_SUBSCRIPTION_SUCCESS_URL,
+)
+from sqlalchemy import select
+
+logger = logging.getLogger(__name__)
+
+
+class AssociationRepository(AbstractRepository):
+
+    # READ
+    async def get_subscription_by_stripe_id(
+        self, stripe_id: str
+    ) -> Optional[Subscription]:
+        query = select(Subscription).where(Subscription.stripe_id == stripe_id)
+        subscription = (await self.session.execute(query)).scalar_one_or_none()
+        return subscription
+
+    async def get_subscription_by_session_id(
+        self, session_id: str
+    ) -> Optional[Subscription]:
+        query = select(Subscription).where(Subscription.stripe_session_id == session_id)
+        subscription = (await self.session.execute(query)).scalar_one_or_none()
+        return subscription
+
+    async def get_subscription_by_customer_id(
+        self, customer_id: str
+    ) -> Optional[Subscription]:
+        query = select(Subscription).where(
+            Subscription.stripe_customer_id == customer_id
+        )
+        subscription = (await self.session.execute(query)).scalar_one_or_none()
+        return subscription
+
+    async def get_subscription_by_user_id(self, user_id: int) -> Optional[Subscription]:
+        query = select(Subscription).where(Subscription.user_id == user_id)
+        subscription = (await self.session.execute(query)).scalar_one_or_none()
+        return subscription
+
+    # WRITE
+    async def save_subscription(self, subscription: Subscription) -> Subscription:
+        self.session.add(subscription)
+        await self.session.flush()
+        return subscription
+
+    # ============== #
+    #    Stripe
+    # ============== #
+    # WRITE TO STRIPE
+    async def create_checkout_session(
+        self, data: StripeCheckoutSessionInput
+    ) -> Optional[StripeCheckoutSession]:
+        # See https://stripe.com/docs/api/checkout/sessions/create
+        # for additional parameters to pass.
+        # {CHECKOUT_SESSION_ID} is a string literal; do not change it!
+        # the actual Session ID is returned in the query parameter when your customer
+        # is redirected to the success page.
+        customer_payload = {}
+        if data.customer_id:
+            customer_payload.update(dict(customer=data.customer_id))
+        elif data.customer_email:
+            customer_payload.update(dict(customer_email=data.customer_email))
+        checkout_session_stripe_key = "{CHECKOUT_SESSION_ID}"
```

Comment:
I am not sure I understand this `{CHECKOUT_SESSION_ID}` here, I saw your comment above but it's still not clear to me 🤔

Reply:
it isis a string literal, a fixed key used by stripe as variable

[Here](https://stripe.com/docs/billing/subscriptions/checkout) maybe it is explained better

Reply:
ok I see now, that's where Stripe puts the session id before redirecting, thanks!